embassy-sync/src/waitqueue/atomic_waker.rs

```diff
@@ -33,9 +33,8 @@ impl<M: RawMutex> GenericAtomicWaker<M> {
     /// Wake the registered waker, if any.
     pub fn wake(&self) {
         self.waker.lock(|cell| {
-            if let Some(w) = cell.replace(None) {
-                w.wake_by_ref();
-                cell.set(Some(w));
+            if let Some(w) = cell.take() {
+                w.wake();
             }
         })
```

Comment:
Would I be able to sell you on this now?
```Rust
        if let Some(w) = self.waker.lock(|cell| cell.take())  {
            w.wake();
        }
```

I'm thinking the lock should be released earlier. You never know how expensive/slow `wake()` might be.